functions.md

```diff
@@ -0,0 +1,302 @@
+
+Furthermore, we can make things even more ergonomic for the user by having a convention that functions which would typically refer to a path segment can also be applied to a path or sketch (by taking advantage of overloading using `self`). For example, `start` might return the start point of a line or arc, and on a path returns the start point of the first segment. Likewise for `end` and the end point of the last segment. `tangent` might refer to the end angle of the last segment of the path, and so forth.
+
+[^self2]: This may remind you of method call syntax in other languages using `.` or `->`. It sort-of is. It also follows `|>` syntax from F#. Note that we use `.` in KCL for field access, including accessing functions if functions are stored in a field (in particular this happens with imported [modules](modules.md)). Note though, that using `.` just locates the function, it does not treat the lhs as the receiver/`self` argument.
+
+[^yikes]: Yikes! This seems a bit subtle and error-prone. However, it does mean that the following examples work: `path |> line(angle = perpendicular(), len = segmentLen())` (`perpendicular` and `segmentLen` are called with `self = path`) and `path |> mirror(line(from = (0, 0), to = end()))` (the non-self version of `line` is used, `path` is passed as `self` to `end`). To get other behaviour, the user can always break the pipeline and use a variable.
+
+### Block syntax
+
+If a function has a `self` argument, then it can be called using a block syntax: `fn-call ::= ... | id '(' namedArg* ')' '{' stmt* '}'`. This is syntax sugar for having the block in `self` position, e.g., `foo(x = 42) { ... }` is equivalent to `foo({ ... }, x = 42)`. Note that the value of the block is assumed to be the expression used on the last line of the block.
+
+The motivation for this is as an alternate syntax for creating sketches which does not emphasise ordering and which better supports constraints and construction geometry, e.g.,
+
+```
+sketch(on = XY) {
+  l1 = line(...)
+  l2 = line(...)
+  cnstCircle = circle(...)
+  cnstLine(...)
+  l3 = line(...)
+  enclose(l1, l2, l3)
+}
+```
+
+Here, having the block inline would be less readable.
+
+## Naming and overriding
+
+Function declarations must have unique names. E.g.,
+
+```
+fn foo() {...}      // ok
+fn bar() {...}      // ok
+fn bar(): num {...} // error: duplicate names
+```
+
+Exception: functions may have the same name if they have fully distinct `self` types. A function with no self argument is considered distinct to any functions with `self`. If the `self` types are unrelated, there is no restriction on the other arguments or return type. If the `self` types are subtypes, other arguments must match exactly[^contra] and return types must be covariant. Due to the `self`'array sugar, `T` is not considered distinct from `[T]` or `[T; 1]`.
```

Comment:
What's the motivation for this? Seems like this would make our hover and autocomplete support very difficult.

Reply:
It should work with hover and autocomplete since it can all be done statically. Not sure what your referring to exactly with 'this', but the general concept is to support virtual dispatch, e.g., in the example how `line` is implemented as a free function and on sketch (and we might want to implement it differently on different kinds of sketch)